### DISCUSSION TITLE
Miserable performance on production

### DISCUSSION THREAD

**Discussion by gauravraj-v** [saleor/saleor#10264] [Deployment]
We started using Saleor 3.1 and added some custom features on top if it related to our e-commerce. Now we deployed it on production with 4 - D4 machines on Azure. They are deployed using docker as suggested on the website. We are using PostgreSQL hosted solution as the backend. We're trying to load test the checkout and orders flow and each of the mutations are taking around 300-500ms. While increasing concurrent users to 10, the latencies start increasing. With 20 users making concurrent requests, latencies have gone to 5-10 seconds. The mutations being run are like Checkout create, shipping method update, delivery method update, item removal, and item addition. The overall QPS of the system is hardly above 23. 
These latencies don't make sense at all for such low load. What could be the reasons for this? And are there any configurations which we should be looking at. 

Added the details of this issue on stackoverflow as well - https://stackoverflow.com/questions/73187360/django-application-with-asgi-uvicorn-increasing-the-latencies-by-4x

**Comment by vamshiaruru** [comment]
+1
In general what are the best practices for deploying saleor to production?

**Reply by peelar** [reply]
You can expect us to add some more content regarding Saleor deployment in the next couple of weeks. I will keep you posted 😉.

**Reply by lack-of-gravitas** [reply]
Is this now available for public consumption? Comment says docs in the next couple of weeks?

**Comment by manojLondhe** [comment]
You may find few pointers in https://github.com/saleor/saleor/discussions/7172 by other users (not from core Saleor team)

**Reply by gauravraj-v** [reply]
Tried these pointers. They mostly talk about machine size and memory utilization. We have 4 dedicated machines with 8GB rams and 4 cores just running Saleor. Still seeing these high latencies.

**Comment by timuric** [comment]
@gauravraj-v could you please share some details about customization, do they affect any of the mutations you have mentioned? Are you using any sync webhooks related to checkout?

**Reply by gauravraj-v** [reply]
No. The customizations are completely independent of the core saleor. They are independent entities and mutations around it.  We are only using async webhooks.

**Reply by timuric** [reply]
👍

**Comment by peelar** [comment]
@gauravraj-v 

Thank you for reaching out.

It appears the latencies don't occur at the API level, but rather somewhere higher. Given your current workload of 4 instances with 4 workers each, you should be able to handle 16 concurrent requests. In your case, however, the problems begin with 10.

Our hypothesis is that your load balancing config causes the traffic to go to a single instance (with 4 workers), instead of splitting it evenly between instances.

Also, another suggestion from our side is increasing the number of workers. With 16GB of memory, you should be able to run 20.

Please let us know if that solved it for you.

**Reply by gauravraj-v** [reply]
Actually we tried on 16GB of RAM with 4 core machines. The latencies don't seem to be improving much. 
And no, the ELB is perfectly distributing load among all the machines. The CPU usage is exactly 50-60% on all the instances.

We checked the traces properly for each requests and noticed that CheckoutCreate mutation and shippingMethodUpdate mutations are taking approximately 300-500ms consistently. The extra delay happening because of ASGI and Uvicorn layer. Then we deployed 8 docker instances on each machine and added nginx to each instance to switch between the dockers. This setup was giving us much better results compared to the recommended setup using ASGI + Uvicorn/Gunicorn.

**Reply by gauravraj-v** [reply]
Still trying to figure out the production setup to get the best performance and support expected number of concurrent requests.
